collab-document/src/blocks/block.rs

```diff
@@ -1,158 +1,160 @@
+    let data = match data {
+      BlockDataEnum::Page(text) | BlockDataEnum::Text(text) => serde_json::json!({
+        "text": text,
+      }),
+      BlockDataEnum::Header(level, text) => serde_json::json!({
+        "level": level,
+        "text": text,
+      }),
+      _ => serde_json::json!({}),
+    };
+    s.serialize_field("data", &data)?;
+
+    s.end()
   }
 }
-
 pub struct BlockMap {
   root: MapRefWrapper,
 }
-impl BlockMap {
-  pub fn new(root: MapRefWrapper) -> Self {
-    Self { root }
-  }
 
-  pub fn get_block(&self, block_id: &str) -> Option<BlockMapRef> {
-    let txn = self.root.transact();
-    let map_ref = self.root.get_map_with_txn(&txn, block_id)?;
-    let block_map = BlockMapRef::from_map_ref(map_ref);
-    drop(txn);
-    Some(block_map)
-  }
-
-  pub fn create_block<B>(&self, block_id: &str, f: B)
+impl Serialize for BlockMap {
+  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
   where
-    B: FnOnce(BlockBuilder) -> BlockMapRef,
+    S: Serializer,
   {
-    self.root.with_transact_mut(|txn| {
-      let builder = BlockBuilder::new_with_txn(txn, block_id.to_string(), &self.root);
-      let _ = f(builder);
-    })
-  }
-
-  pub fn insert_block(&self, block: Block) {
-    self.root.with_transact_mut(|txn| {
-      self
-        .root
-        .insert_json_with_txn(txn, &block.id.clone(), block)
-    })
-  }
-}
-
-impl Deref for BlockMap {
-  type Target = MapRefWrapper;
-
-  fn deref(&self) -> &Self::Target {
-    &self.root
+    let txn = self.root.transact();
+    let mut map = serializer.serialize_map(Some(self.root.len(&txn) as usize))?;
+    for (k, _) in self.root.iter(&txn) {
+      let block = self.get_block(&txn, k).unwrap();
+      let value = serde_json::to_value(block).unwrap();
+      map.serialize_entry(k, &value)?;
+    }
+    map.end()
   }
 }
 
-pub struct BlockBuilder<'a, 'b> {
-  block_map: BlockMapRef,
-  txn: &'a mut TransactionMut<'b>,
-}
-
-impl<'a, 'b> BlockBuilder<'a, 'b> {
-  pub fn new(txn: &'a mut TransactionMut<'b>, container: &MapRefWrapper) -> Self {
-    let key = nanoid!(4);
-    Self::new_with_txn(txn, key, container)
+impl BlockMap {
+  pub fn new(root: MapRefWrapper) -> Self {
+    Self { root }
   }
 
-  pub fn new_with_txn(
-    txn: &'a mut TransactionMut<'b>,
-    block_id: String,
-    container: &MapRefWrapper,
-  ) -> Self {
-    let map_ref = match container.get_map_with_txn(txn, &block_id) {
-      None => container.insert_map_with_txn(txn, &block_id),
-      Some(map) => map,
-    };
-    let block_map = BlockMapRef::from_map_ref(map_ref);
-
-    Self { block_map, txn }
+  pub fn to_json(&self) -> serde_json::Value {
+    serde_json::to_value(self).unwrap()
   }
 
-  pub fn with_type<T: AsRef<str>>(mut self, ty: T) -> Self {
-    self.block_map.set_ty(self.txn, ty.as_ref().to_string());
-    self
+  pub fn get_block<T: ReadTxn>(&self, txn: &T, block_id: &str) -> Option<Block> {
+    let block_map = self.root.get_map_with_txn(txn, block_id);
```

Comment:
try `and_then` or `map`. for example:

self.root.get_map_with_txn().and_then()
self.root.get_map_with_txn().map()

Reply:
How about this?

```rs
  pub fn get_block<T: ReadTxn>(&self, txn: &T, block_id: &str) -> Option<Block> {
    let block_map = self.root.get_map_with_txn(txn, block_id)?;
    self.get_block_by_map(txn, block_map)
  }

  pub fn get_block_by_map<T: ReadTxn>(&self, txn: &T, block_map: MapRefWrapper) -> Option<Block> {
    let id = block_map.get_str_with_txn(txn, "id")?;
    let ty = block_map.get_str_with_txn(txn, "ty")?;
    let parent = block_map.get_str_with_txn(txn, "parent")?;
    let children = block_map.get_str_with_txn(txn, "children")?;
    let data = block_map.get_str_with_txn(txn, "data")?;

    Some(Block {
      id,
      ty,
      parent,
      children,
      data,
    })
  }
```